modules/distributor/forwarder/otlpgrpc/forwarder.go

```diff
@@ -0,0 +1,164 @@
+package otlpgrpc
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/go-kit/log"
+	grpcmw "github.com/grpc-ecosystem/go-grpc-middleware"
+	"github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/weaveworks/common/middleware"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/collector/pdata/ptrace/ptraceotlp"
+	"go.uber.org/multierr"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/grafana/tempo/pkg/tempopb"
+)
+
+type Forwarder struct {
+	cfg         Config
+	logger      log.Logger
+	connections map[string]*grpc.ClientConn
+	clients     map[string]ptraceotlp.Client
+	initialized bool
+	mu          *sync.RWMutex
+}
+
+func NewForwarder(cfg Config, logger log.Logger) (*Forwarder, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, fmt.Errorf("failed to validate config: %w", err)
+	}
+
+	return &Forwarder{
+		cfg:         cfg,
+		logger:      logger,
+		connections: make(map[string]*grpc.ClientConn),
+		clients:     make(map[string]ptraceotlp.Client),
+		initialized: false,
+		mu:          &sync.RWMutex{},
+	}, nil
+}
+
+// Dial creates client connections and clients based on config.
+// Dial is expected to be called only once.
+func (f *Forwarder) Dial(ctx context.Context, opts ...grpc.DialOption) error {
+	f.mu.Lock()
```

Comment:
i would really like to get rid of the locking code in this object. i think we can do that using the concurrency controls in the forwarder manager one package up. basically, making sure that lifecycle events Dial, ForwardBatches, and Shutdown are never called concurrently. it may already be true.

Reply:
I wouldn't change this. The idea is that many tenants can "attach" to a single Forwarder instance. The flow is kind of like this:

Distributor -> Queue (per tenant) -> Forwarder

Many tenants share the same forwarder. Also, the Dial is just called once, during Tempo startup, and `ForwardBatches` only acquires a read-lock and does so on Queue's thread (in workers, essentially). I don't think we have much to gain with this micro-optimization.

Reply:
For me, it's less about optimization and more about only using locks where necessary to reduce code complexity. However, I consider this minor am fine with leaving it in.